server/src/main/java/org/opensearch/index/store/remote/utils/TransferManager.java

```diff
@@ -50,63 +48,115 @@ public TransferManager(final BlobContainer blobContainer, final FileCache fileCa
-                    } catch (IOException e) {
-                        logger.warn("Failed to download " + blobFetchRequest.getFilePath(), e);
-                        return null;
+        return AccessController.doPrivileged((PrivilegedAction<FileCachedIndexInput>) () -> {
+            try {
+                if (Files.exists(request.getFilePath()) == false) {
+                    try (
+                        InputStream snapshotFileInputStream = blobContainer.readBlob(
+                            request.getBlobName(),
+                            request.getPosition(),
+                            request.getLength()
+                        );
+                        OutputStream fileOutputStream = Files.newOutputStream(request.getFilePath());
+                        OutputStream localFileOutputStream = new BufferedOutputStream(fileOutputStream)
+                    ) {
+                        snapshotFileInputStream.transferTo(localFileOutputStream);
                     }
-                } else {
-                    if (cachedIndexInput.isClosed()) {
-                        // if it's already in the file cache, but closed, open it and replace the original one
-                        try {
-                            final IndexInput luceneIndexInput = blobFetchRequest.getDirectory()
-                                .openInput(blobFetchRequest.getFileName(), IOContext.READ);
-                            return new FileCachedIndexInput(fileCache, blobFetchRequest.getFilePath(), luceneIndexInput);
-                        } catch (IOException e) {
-                            logger.warn("Failed to open existing file for " + blobFetchRequest.getFilePath(), e);
-                            return null;
-                        }
-                    }
-                    // already in the cache and ready to be used (open)
-                    return cachedIndexInput;
                 }
-            })
-        );
+                final IndexInput luceneIndexInput = request.getDirectory().openInput(request.getFileName(), IOContext.READ);
+                return new FileCachedIndexInput(fileCache, request.getFilePath(), luceneIndexInput);
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        });
+    }
+
+    /**
+     * Implementation of CachedIndexInput the defers creation of the underlying
+     * IndexInput until the first invocation of {@link #getIndexInput()}. This
+     * class is thread safe, and concurrent calls to {@link #getIndexInput()} will
+     * result in blocking until the initial thread completes the creation of the
+     * IndexInput.
+     */
+    private static class DelayedCreationCachedIndexInput implements CachedIndexInput {
+        private final FileCache fileCache;
+        private final BlobContainer blobContainer;
+        private final BlobFetchRequest request;
+        private volatile boolean isClosed = false;
+        private Object result;
```

Comment:
I want to hold either an IndexInput or a RuntimeException here, but never both. This approach, while it looses type safety and requires casts, does enforce that you cannot have both values. I could also implement it using two values and implement the logic to set one or the other. I'm open to other suggestions as well.

Reply:
If multiple requests are blocking on one remote fetch, I want all of those threads to get a reference to the exception (if it fails) so they can propagate it up, so I only need it long enough to share between threads, but I do need to store it for that reason. Ultimately I think it is better to propagate the error and let retries happen higher up, rather than having threads queue up here and implicitly retry, potentially all running into the same failure.

I considered doing another internal type to encapsulate the types, but it kills me how verbose it ends up being:

```
private static abstract class Result {
    RuntimeException error() {
        throw new UnsupportedOperationException();
    }
    IndexInput indexInput() {
        throw new UnsupportedOperationException();
    }
    abstract boolean isError();

    private static class Success extends Result {
        private final IndexInput indexInput;

        private Success(IndexInput indexInput) {
            this.indexInput = indexInput;
        }

        @Override
        IndexInput indexInput() {
            return indexInput;
        }

        @Override
        boolean isError() {
            return false;
        }
    }

    private static class Error extends Result {
        private final RuntimeException error;

        private Error(RuntimeException error) {
            this.error = error;
        }

        @Override
        RuntimeException error() {
            return error;
        }

        @Override
        boolean isError() {
            return true;
        }
    }
}
```

What do you think?